src/app/clusters/soil-measurement-server/soil-measurement-server.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "soil-measurement-server.h"
+
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <zap-generated/gen_config.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace SoilMeasurement {
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::SoilMeasurement;
+using namespace chip::app::Clusters::SoilMeasurement::Attributes;
+
+MeasurementData gMeasurements[MATTER_DM_SOIL_MEASUREMENT_CLUSTER_SERVER_ENDPOINT_COUNT + CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT];
+
+Globals::Structs::MeasurementAccuracyRangeStruct::Type soilMoistureMeasurementLimitsAccuracyRange[] = {
+    { .rangeMin = 0, .rangeMax = 100, .percentMax = MakeOptional(static_cast<chip::Percent100ths>(10)) }
+};
+
+CHIP_ERROR SoilMeasurementAttrAccess::Init()
+{
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
+
+    // Initialize the soil moisture measurement limits to default values
+    for (auto & measurement : gMeasurements)
+    {
+        measurement.soilMoistureMeasurementLimits = {
+            .measurementType  = Globals::MeasurementTypeEnum::kSoilMoisture,
+            .measured         = true,
+            .minMeasuredValue = 0,
+            .maxMeasuredValue = 100,
+            .accuracyRanges   = DataModel::List<const Globals::Structs::MeasurementAccuracyRangeStruct::Type>(
+                soilMoistureMeasurementLimitsAccuracyRange)
+        };
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void SoilMeasurementAttrAccess::Shutdown()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+}
+
+CHIP_ERROR SoilMeasurementAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == SoilMeasurement::Id);
+
+    MeasurementData * data = SoilMeasurementDataForEndpoint(aPath.mEndpointId);
```

Comment:
Is this used anywhere?

Reply:
It was used by the read of the limits. I have modified this to move this into the reads which requires it.